### DISCUSSION TITLE
Novelibre and MaweJS?

### DISCUSSION THREAD

**Discussion by mkoskim** [peter88213/novelibre#32] [General]
Hi!

I am developing my own writer's editor named MaweJS:

https://github.com/mkoskim/mawejs

I looked your project, and took a look to file format: Novelibre is using XML just like MaweJS does. If you have any interests, you could help me to import Novelibre XML files to my editor, and I could help importing MaweJS files to Novelibre.

What I am looking for on my own editor, is the possibility to keep native formats as long as possible:

https://github.com/mkoskim/mawejs/issues/153

- - -

To get some idea of MaweJS, there are two Wiki pages:

1) General introduction: https://github.com/mkoskim/mawejs/blob/master/README.md

2) Short history: https://github.com/mkoskim/mawejs/wiki/Short-history-of-MaweJS

**Comment by peter88213** [comment]
Hi Markus, 

your MaweJS word processor looks great at the first glance. If I hadn't got novelibre, I certainly would give it a try. 

As far as the novx file format is concerned, you have probably already seen the [documentation](https://peter88213.github.io/novxlib-docs/the_novx_file_format.html). 
I have created [a separate repository](https://github.com/peter88213/novxlib) with a Python library that reads and writes the novx files and provides an object-oriented data model that can be used to realise converters to any other file format. The novxlib library provides classes for the export and import of ODT and ODS documents. 

The whole thing derives from my earlier work with the yWriter file format; there is a [broad ecosystem](https://github.com/peter88213/pywriter) for connecting yWriter to a variety of other applications, including my own writing programme [novelyst](https://github.com/peter88213/novelibre/discussions/2), the predecessor of novelibre. 

I read your article about the different file formats; as far as yWriter is concerned, with yw7 there is a single file format available that is very flexible and extensible. It's XML, and the scene contents are integrated as CDATA areas with yWriter's native shortcode markup.  It is undocumented and is quietly extended from time to time, but it is easy to figure out and designed for backwards compatibility. 

My novx format is based on the yw7 format, but is XML throughout, making it easier to parse and validate. The text formatting is largely done with XHTML tags; only for comments and footnotes I have defined my own XML elements, which are simplifications of the ODF elements. I have also introduced elements for plot lines, plot points and plot notes, which could only be implemented as workarounds by convention with yw7. The disadvantage of novx compared to yw7, if you want to see it as a disadvantage, is the [limited number of formatting tags](https://peter88213.github.io/nvhelp-en/basic_concepts.html#formatting-text). I have decided to support only the formatting that is essential for a prose manuscript. 

If you are still looking for a file format for MaweJS, it would be worth taking a second look at yw7, especially since, as I said, there are a large number of conversion and synchronisation tools. 

Then there's Manuskript with its file system-based data format. I'm not very impressed, however, I once wrote a [document exporter](https://github.com/peter88213/manuskript-exporter/), just for fun. 

Another file system-based format with some distribution has [novelWriter](https://novelwriter.io/docs/). I programmed a yw7 converter for this as well. But yet well-documented, the format is complicated to process, I don't have much love for it. 

Finally, something else: as a sporting exercise, I am creating [mdnovel](https://github.com/peter88213/mdnovel), a variant of novelibre that works entirely with Markdown and has an integrated text editor. The downloadable version still reads and writes the novx format. In the development branch, however, I am working with a new file format based on YAML and Markdown, very similar to Obsidian.

**Reply by mkoskim** [reply]
Yeah, I might first make the file format more compatible to yw7 / novelibre, because at the moment I have all the freedom do whatever I like with the file format. Making it first more compatible - eg. doing format version 3 - should ease the interoperability.

Just quick note: Because MaweJS is internally a structured editor, it of course uses structured file format, and thus it should not be too hard to tailor the format more suitable for other editors.

**Reply by peter88213** [reply]
Before I programmed my own writing application, I spent years creating file converters for the yWriter format. That's why the core of my program is the data model. There is a *Novel* class, which contains the novel text, the metadata and the structural information. Then there is the abstract *File* class, which serves as an interface for read and write operations and for converting the text formatting. The *File* class delegates a *Novel* instance that is passed between different *File* subclass instances that are tailored to the different file formats.  

This may sound complicated, but it allows the application to be perfectly decoupled from the file format. As already mentioned, I am currently practicing this on an extreme example where XML is replaced by Markdown/YAML.

**Reply by mkoskim** [reply]
Yes, I use pretty similar decoupling (in all my editors, moe, Python mawe and MaweJS). While I am currently using XML in the files, they could be quite easily changed to JSON, or gather the tree from multiple files, just by writing load & save functions:

https://github.com/mkoskim/mawejs/tree/v0.4.0/src/document/xmljs
